### ISSUE TITLE
tgl_timer_methods.insert called too often

### ISSUE THREAD

**Issue by BenWiederhake** [vysheng/tgl#49]
While investigating a new bug, I noticed that tgl calls `.insert` before the old timer happened, and without calling `.remove` first.

Here's a backtrace: http://pastebin.com/ihdTvwfc

Here's my summary:

- `tgl_timer_insert` (in telegram-purple), called by:
- `tree_act_ex_query` (in queries.c:113), called by:
- `tree_act_ex_query` (in queries.c:113), called by:
- `tree_act_ex_query` (in queries.c:113), called by:
- `tree_act_ex_query` (in queries.c:113), called by:
- `tglq_regen_queries_from_old_session` (in queries.c:199), called by:
- `rpc_execute_answer` (in mtproto-client.c:831), called by:
- `rpc_execute_answer` (in mtproto-client.c:831), called by:
- `rpc_execute_answer` (in mtproto-client.c:831), called by:
- `rpc_execute_answer` (in mtproto-client.c:831), called by:
- `rpc_execute` (in mtproto-client.c:1148)

I think this is a impendence mismatch in `tree_act_ex_query`: It calls *itself* recursively, which calls `tglq_regen_queries_from_old_session` (given as argument), which *again* calls `tree_act_ex_query` recursively.

**Comment by majn** [comment]
Calling tgl_timer_insert multiple times shouldn't be an issue.

What the function appears to do is to iterate through all running queries to restart them. The reason why the function shows up multiple times is that tree.h implements it as a recursive function (https://github.com/vysheng/tgl/blob/master/tree.h#L133-L147).

**Comment by majn** [comment]
There's something off about this. The issue appears to be caused by an assertion assert (-1 == t->fd), which is an assertion I'm not able to find anywhere in the code.

**Comment by BenWiederhake** [comment]
Of course it is an issue:

- instantiation
- call with insert, internal fd is set to <whatever>
- another call with insert, internal fd is set to <whatother>
- call to free, internal freeing stops the <whatother> timer, but NOT the <whatever> one.
- <whatever> timer fires, doing a use-after-free

**Comment by BenWiederhake** [comment]
That assertion is written by me.

I'm making the assumption that when insert is called, there should be no other timer currently being "ticking"

**Comment by majn** [comment]
You are right, of course it shouldn't be called multiple times on the same timer_alarm struct.

**Comment by majn** [comment]
I've debugged this issue and printed a stacktrace for every time timer insert is called. I've added the same assertion as Ben.

https://gist.github.com/majn/0c69dd738d7690222489

Apparently, tgl_do_get_difference is executed for the first time and starts the regular alarm timeout of (6 seconds) and then when work_new_session_created is received, the second timeout (0.001s) is called on the _same_ event struct for the second time.

Is this behavior by design and the timer implementation is supposed to cancel all pending timeouts on its own?

**Comment by majn** [comment]
One more thing:

work_new_session_created appears to happen every time the tgl is started, restarting all running queries on every restart.

**Comment by BenWiederhake** [comment]
From @vysheng: The intended meaning of that was precisely like we assumed.
